polaris/ocean/ice_shelf/ssh_adjustment.py

```diff
@@ -0,0 +1,150 @@
+from mpas_tools.cime.constants import constants
+from mpas_tools.io import write_netcdf
+
+from polaris import Step
+
+
+class SshAdjustment(Step):
+    """
+    A step for iteratively adjusting the pressure from the weight of the ice
+    shelf to match the sea-surface height as part of ice-shelf 2D test cases
+    """
+    def __init__(self, component, resolution, init, forward, indir=None,
+                 name='ssh_adjust'):
+        """
+        Create the step
+
+        Parameters
+        ----------
+        resolution : float
+            The resolution of the test case in m
+
+        coord_type: str
+            The coordinate type (e.g., 'z-star', 'single_layer', etc.)
+
+        iteration : int, optional
+            the iteration number
+        """
+        self.resolution = resolution
+
+        super().__init__(component=component, name=name,
+                         indir=f'{indir}/ssh_adjustment')
+
+        self.add_input_file(filename='final.nc',
+                            work_dir_target=f'{forward.path}/output.nc')
+        self.add_input_file(filename='init.nc',
+                            work_dir_target=f'{forward.path}/init.nc')
+        self.add_input_file(filename='mesh.nc',
+                            work_dir_target=f'{init.path}/culled_mesh.nc')
+        self.add_output_file(filename='output.nc')
+
+    # no setup() is needed
+
+    def run(self):
+        """
+        Adjust the sea surface height or land-ice pressure to be dynamically
+        consistent with one another.
+
+        """
+        logger = self.logger
+        config = self.config
+        adjust_variable = 'landIcePressure'
+        final_filename = self.inputs[0]
+        init_filename = self.inputs[1]
+        mesh_filename = self.inputs[2]
+        out_filename = self.outputs[0]
+        ds_mesh = xr.open_dataset(mesh_filename)
+        ds_init = xr.open_dataset(init_filename)
+        ds_final = xr.open_dataset(final_filename)
+        ds_out = ds_init.copy()
+
+        if adjust_variable not in ['ssh', 'landIcePressure']:
+            raise ValueError(f"Unknown variable to modify: {adjust_variable}")
+
+        logger.info("   * Updating SSH or land-ice pressure")
+
+        # keep the data set with Time for output
+        # and generate these time slices
+        ds_init = ds_init.isel(Time=0)
+        ds_final = ds_final.isel(Time=-1)
+
+        on_a_sphere = ds_out.attrs['on_a_sphere'].lower() == 'yes'
+
+        if 'minLevelCell' in ds_final:
+            minLevelCell = ds_final.minLevelCell - 1
+        else:
+            minLevelCell = ds_mesh.minLevelCell - 1
+
+        init_ssh = ds_init.ssh
+        final_ssh = ds_final.ssh
+        top_density = ds_final.density.isel(nVertLevels=minLevelCell)
+
+        y3 = config.getfloat('ice_shelf_2d', 'y3') * 1e3
+        mask = np.logical_and(ds_final.maxLevelCell > 0,
+                              ds_mesh.yCell < y3).astype(float)
+        delta_ssh = mask * (final_ssh - init_ssh)
```

Comment:
This is specific to `ice_shelf_2d` and needs to be handled in a different way somehow.  Presumably a mask for this purpose needs to be stored in `init.nc`.

Reply:
@xylar Thanks for pointing this out. I agree that we should store it in `init.nc`. Do you think there's a mask that we would want to use for all ice shelf cases, perhaps `landIceFraction > 0`? Seems preferable to use an existing land ice variable.

Reply:
I think it would be best to require a new variable for this mask.  It might be useful to adjust the SSH even outside of cavities (e.g. if topography smoothing has been performed).

MPAS-Ocean init mode has `modifyLandIcePressureMask`:
https://github.com/E3SM-Project/E3SM/blob/f427d86d53cce477a43af446a898e14ece0d0b4e/components/mpas-ocean/src/Registry.xml#L2801-L2804
but I think we should probably call it `modifySSHMask` or something like that.

Reply:
I added a config option called `mask_variable` to designate that variable's name. I called it `adjustSSHMask` by default, to try to be intuitively relatable to `ssh_adjustment`